src/Polysemy/Meta.hs

```diff
@@ -0,0 +1,300 @@
+{-# language AllowAmbiguousTypes, BangPatterns, GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE FlexibleInstances, FunctionalDependencies, UndecidableInstances #-}
+
+-- | Description: Interpreters for 'Meta'
+module Polysemy.Meta (
+  -- * Effect
+  Meta,
+  MetaEffect,
+  (:%),
+  mkIntoMeta,
+
+  -- * Actions
+  sendMeta,
+  sendMetaUsing,
+
+  -- ** Last-resort actions
+  sendMetaVia,
+  sendMetaViaUsing,
+
+  -- * Interpreters
+  interpretMeta,
+  reinterpretMeta,
+  reinterpretMeta2,
+  reinterpretMeta3,
+  metaToMeta,
+  metaToMetaUsing,
+  metaIntoMeta,
+
+  -- * 'interpretMeta' Combinators
+  HandlingMeta,
+  runMeta,
+  runMeta',
+  runExposeMeta,
+  runExposeMeta',
+
+  -- ** Last-resort 'interpretMeta' combinators
+  runMetaUsing,
+  runMetaUsing',
+  runExposeMetaUsing,
+  runExposeMetaUsing',
+
+  -- * Auxiliary type classes
+  DepMember,
+  AllSemRaises,
+  ) where
+
+import Data.Unique
+import Control.Monad
+
+import Polysemy
+import Polysemy.Bundle
+import Polysemy.Membership
+import Polysemy.Internal.Utils
+import Polysemy.Internal.Meta
+import Polysemy.Internal.HigherOrder
+import System.IO.Unsafe
+import Unsafe.Coerce
+
+data HandlingMeta metaeffect t rH l :: Effect where
+  HandlingMetaMetaRun
+    :: forall metaeffect t rH l m a
+     . MetaRun m a -> HandlingMeta metaeffect t rH l m a
+  ProcessMeta
+    :: forall eff z metaeffect t rH l m a
+     . Unique
+    -> ElemOf '(z, eff) l
+    -> t ()
+    -> z a
+    -> HandlingMeta metaeffect t rH l m
+        (Sem (Meta metaeffect ': HandlingMeta metaeffect t rH l ': rH) (t a))
+
+data Box where
```

Comment:
Is the value in the box ever used?

Reply:
So `Box` is just what's used to carry an argument to `newUnique'`. `newUnique'` has an argument to ensure that the `unsafePerformIO (newUnique' ...)` _absolutely,  definitively_ does not get floated out of `processMeta`. If we used `newUnique` instead:
```haskell
processMeta pr t z = do
    let !uniq = unsafePerformIO newUnique
    ....
```
then GHC is free to transform this to:
```haskell
_newUnique :: Unique
_newUnique = unsafePerformIO newUnique

processMeta pr t z = do
    let !uniq = _newUnique
    ....
```
which is bad. the `NOINLINE` to `newUnique'` is so that it doesn't inline `newUnique' (Box z)`, to `newUnique` and is _then_ free to do the above.

Finally, providing and pattern-matching on `Box` instead of a generic, unevaluated `a` is just a last, extremely paranoid measure on my part: it avoids a theoretical GHC optimization that would detect if a function is always lazy in its arguments, and then rewrite `newUnique'` to:
```haskell
_newUnique' :: IO Unique
_newUnique' = newUnique
{-# NOINLINE newUnique'_ #-}

newUnique' :: a -> IO Unique
newUnique' _ = _newUnique'
{-# INLINE newUnique' #-}
```
... because then `unsafePerformIO (newUnique' z)`  would get inlined to `unsafePerformIO _newUnique', and then can get floated out.

Reply:
did you observe it being floated out without those measures?